### ISSUE TITLE
Joins

### ISSUE THREAD

**Issue by techfort** [techfort/LokiJS#106]
At the moment Loki supports in-memory joins e.g.
```javascript
var author = authors.insert({name: 'Stephen King'});
var book = books.insert({title: 'Salems Lot', author: author})
```
But this will not persist correctly as the author will be serialized as an embedded object, not the original author object. There should be a more solid mechanism to operate joins.

**Comment by jrhicks** [comment]
Assuming that we have external ids

    var author = authors.insert({id: 1, name: 'Stephen King'});
    var book = books.insert({title: 'Salems Lot', author: 1})

Do I join like?

    function bookJoin(obj){
        obj.author = author.findOne({id: obj.author})
    }

    function noReduction(array) {
        return array;
    }

    books.mapReduce(bookJoin, noReduction);

My dream API would probably look something like:

    var bookView = books.addDynamicView('bookView');
    bookView.applyJoin( {author: function(obj) { return author.findOne({id: obj.author}) });

    bookView.data();

Or more generally maybe I could apply a map to a dynamic view

    var bookView = books.addDynamicView('bookView');
    bookView.applyMap( {author: function(obj) { return objectAssign(obj, author: author.findOne({id: obj.author}) } );

**Comment by techfort** [comment]
Thanks Rob, very interesting! It all makes a lot of sense and the API you used is also very consistent with the existing one. Let's get @obeliskos  involved too and see what we can come up with.

**Comment by assafweinberg** [comment]
A couple months ago, I converted a big project from websql to Lokijs. I had to rewrite about 800 complex relational queries into loki's fluent syntax.  To make things easier, I wrapped a Loki instance in an object that provided a relational friendly interface including joins.

Here's a gist of what the innerJoin function looked like.  
https://gist.github.com/assafweinberg/4c7ef4e833d0493c693b

It worked pretty well and was super flexible. But it wasn't chainable. It just returned an array of cloned objects.

I'd love to figure out how we can add a simple join to the resultset. 

e.g. (roughly)

```
books.find({year: 2008})
    .join('authors', 'authorId', 'id')
    .join('publishers', 'publisherId', 'id')
    .data();

//or

books.find({year: 2008})
    .join('authors', function(book,author){return book.authorId == author.id})
    .filter(function(obj){...})
    .data();
```

Aside from the API, we should also think about what structure the resulting objects should be and how to deal with name collisions.  RethinkDB (http://rethinkdb.com/docs/table-joins/) takes a pretty conservative approach and adds a 'zip()' function.  Also, I think the objects returned from a join have to be clones, so they don't affect the original objects in the collection. Not sure if this would be confusing at all.

What do you guys think?

**Comment by techfort** [comment]
I think the gist is awesome and this whole conversation is stellar. I think the most amazing thing is that @jrhicks  is illustrating the adoption of joins in a DynamicView  and @assafweinberg in the Resultset fluent API, the two main querying facilities of Loki. Maybe it's just me but I find getting to the same result from diametrically opposite approaches is fantastic. As far as I'm concerned the API looks great, we'll wait on @obeliskos 's view on this too, but this is great stuff in the meantime!

**Comment by jrhicks** [comment]
@assafweinberg thanks for the gist!  Clones make sense for anything returned from applyMap.  The structure of applyMap would be 100% under the control of the map function.  The structure of join (as you suggest) needs some consensus.  I don't have an opinion there yet :)

**Comment by techfort** [comment]
By the way - yes I agree with @jrhicks that clones are a better choice when creating joins. I also  like that `zip` a lot, very convenient. As for chain-ability, I think I have something in mind for that, like - for example - wrapping the resulting join array into an anonymous collection and take it from there. All theoretical, but it could work.

**Comment by assafweinberg** [comment]
All good points.

Regarding output format - I'm a little intrigued by the 3rd example from my previous comment - the map/reduce inspired join. In practice, I've found that most joins exist only to get at one or two columns from the joined table rather than all of them. If we let the user provide a map function that specifies exactly how the join happens, we're no longer responsible for output format or name collisions.  It also allows users to create calculated fields.  Seems pretty powerful.  What do you guys think?

**Comment by techfort** [comment]
@assafweinberg that is an extremely interesting point you make, and one made with great timing. I have been working lately on a concept which in my mind i called map-as or map-view (drawing from map-reduce). Basically I thought that sometimes you may have objects of a certain type in your collection, but that's not necessarily how you want to view your data in all circumstances. For example, let's say you have a collection of shapes, each shape following this format:
`[{ sides: [3, 4, 5] , color: 'blue'}, { sides: [4, 4, 4, 4], color: 'red' }]`
that's great, but maybe you want to create a dynamic view which also applies some intelligence to your data and instead of outputting objects it outputs something like 
`[{shape: 'blue triangle', area: 6.0}, { shape: 'red square', area: 16.0}]`. This may be only partially relevant to a collection of objects of the same type, but i would say it becomes more relevant when using joins like you illustrated, and if i get you correctly that is exactly where you going with your comment, or do I misunderstand? If so then i'm more than "a little intrigued", i am very intrigued. And regardless, your delegation of mapping and output to the user seems ideal to me. Loki sacrifices sometimes a little friendliness for performance (as you know :D) so it also conforms with the philosophy of the library.
Which - remember - was written by the god of mischief.

**Comment by assafweinberg** [comment]
Haha, glad to know we're playing god here!  

Yes, we're on the same page. A SQL join in functional terms is essentially just a *match* function that defines how to pair up records from two tables,  followed by a *map* function that defines how to combine data from record pairs into a new object. In Javascript, we can actually define those two operations much more expressively than we can in SQL. 

The example I wrote above, it can be more literally expressed as:
```
books.find()
    .match('authors', function(book,author){return book.authorId == author.id})
    .map(function(obj){return _.assign(obj.left, obj.right)})
    .data();

// or to illustrate @tech_fort's case

shapes.find()
    .match('areaFormulas', function(shape,formula){return shape.sides.length === formula.sides})
    .map(function(obj){
        return {shape: obj.left.name, area: obj.right.formula(obj.left.sides)}
     })
    .data();
```

Does that make sense? This is essentially what RQL does.

Because people are so familiar with the bundling of these operations under the "join" concept, I'd probably keep them together with a syntax like:

```
leftCollection.find()
    .mapJoin(
        'rightCollection', 
        function(leftDoc,rightDoc){return bool}, //Match function
        function(leftDoc,rightDoc){return Object}) //Map function
```

What do you think? Are we going out on a limb here?

**Comment by techfort** [comment]
I say it looks great to me, it's a great approach. We could even give it a stab already but I really want to hear what @obeliskos has to say first, and we can then start prototyping this. Joins would definitely be a highlight for 1.3.0, due in April.

**Comment by jrhicks** [comment]
@assafweinberg I'm trying to understand

```
leftCollection.find()
    .mapJoin(
        'rightCollection', 
        function(leftDoc,rightDoc){return bool}, //Match function
        function(leftDoc,rightDoc){return Object}) //Map function
```
Is the speed of this O(n^2) because the Match function must be called on each combination of left and right doc?

**Comment by jrhicks** [comment]
Couch has this concept of a "design document".  A design document describes how to obtain the linking value from each different type of document in the join and then uses this to maintain a heterogenious collection sorted on the key.

```
// How to join in O(nLog(n))
let c1  = books.data.map( (b)=> {key: b.author_id, collection: 'books', doc: b )} )
let c2 = authors.data.map( (a)=> {key: a.id, collection: 'authors', doc: a})
let combined = c1.concat(c2);
combined.sort( (x,y)=> compare(x.key, y.key) );
```

You could then iterate over the combined in order dumping the results like such:

```
[ {
   books: [
                   {id:42, title: 'A Handbook of Norse Mythology', author_id: 1},
                   {id: 43, title: 'The art of mischievous', author_id: 1}
               ],
   authors: [
                   {id: 1, name: 'Karl Mortensen'}
               ]
 },{
   books: [
                   {id: 69, title: 'Logical Chess: Move By Move', author_id: 2},
                   {id: 70, title: 'The Fireside Book of Chess', author_id: 2}
               ],
   authors: [
                   {id: 2, name: 'Irving Chernev'}
               ]
}]
```

**Comment by techfort** [comment]
Yes I like this very much too, and the reduction in the number of
operations is a definite highlight given our emphasis on performance! How
would you see this implemented respecting the api you illustrated in your
first post (which i really liked)?

**Comment by obeliskos** [comment]
Wow... might take a while to catch up with this convo :)   I actually rather enjoy watching you guys work this out.  Any of these ways would work, most described above would require users code to initiate orchestrate which many times is probably preferred.  

May end up being we eventually provide both a code driven join/mapjoin approach as well as a data driven join feature whereby it can be persisted (as an object closely resembling the input parameters to the first gist innerjoin function).

I like the idea of supporting chaining by dumping into an anonymous collection.

Performance, yep we need to consider this and probably memory consumption more so for loki.  If we are cloning return values may want to do all map functions first and then allow duplicates of clones in 1-many or many-to-many situations.  We still need to describe how to graft into array of single hierarchical documents... correct? Such as property name for object or array of objects (1-many).

**Comment by assafweinberg** [comment]
@jrhicks - You're right that the match function wouldn't be the most performant thing in the world, though it would probably be good enough in most cases.  I'd recommend running some tests to see how bad the problem would be.  Its nice because it gives users the opportunity to write complex ON clauses.  

If we had to drop it, we could do something like this using the join field names, which would be more performant since it lets us take advantage of indices or create a lookup table on the fly.

```
leftCollection.find()
    .mapJoin(
        'rightCollection', 
        'leftCollectionJoinField', //Join field name from left collection
        'rightCollectionJoinField', //Join field name from right collection
        function(leftDoc,rightDoc){return Object}) //Map function
```

Again, the disadvantage of that syntax is that you can't join on multiple columns or derived values.  We could enable that capability though, in a more performant way, using the design document concept you wrote about earlier.  Basically, creating a design document on the fly in a query.

```
leftCollection.find()
    .mapJoin(
        'rightCollection', 
        function(doc){return leftKeyValue}, //Calculated join field for left documents
        function(doc){return rightKeyValue}, //Calculated join field for right documents
        function(leftDoc,rightDoc){return Object}) //Map function
```

Custom joins are also much less useful than the built in map function, so I wouldn't be opposed to dropping it if we had to.





Thanks for the info on the design document stuff.  That's pretty cool

**Comment by jrhicks** [comment]
This last API example is really cool, flexible, and fast.  I think it is a good example of a result-set-join API.

```
leftCollection.find()
    .mapJoin(
        'rightCollection', 
        function(doc){return leftKeyValue}, //Calculated join field for left documents
        function(doc){return rightKeyValue}, //Calculated join field for right documents
        function(leftDoc,rightDoc){return Object}) //Map function
``
Since the results aren't merely grouped instead of mapped, I may be mistaken, but we may not have to clone with this API.
`
[ {
   books: [
                   {id:42, title: 'A Handbook of Norse Mythology', author_id: 1},
                   {id: 43, title: 'The art of mischievous', author_id: 1}
               ],
   authors: [
                   {id: 1, name: 'Karl Mortensen'}
               ]
 },{
   books: [
                   {id: 69, title: 'Logical Chess: Move By Move', author_id: 2},
                   {id: 70, title: 'The Fireside Book of Chess', author_id: 2}
               ],
   authors: [
                   {id: 2, name: 'Irving Chernev'}
               ]
}]

**Comment by assafweinberg** [comment]
I think you're right - none of the mapJoin patterns actually require cloning objects because the mapJoin returns new objects created in the map function.

Another question this brings up... are we doing inner or left joins or both?

**Comment by techfort** [comment]
@assafweinberg @jrhicks this is really good. 2 things: 
* for joins to be applied to a dynamic-view, the speed of creating the joins doesn't matter that much because the view recomputes itself on db changes. Yes, we can measure performance in terms of ops/s but in the real world this is also measured by how quickly data is made available to other application components. If a view (dynamicview) re-computes itself in the background, then data is immediately available when other components need it, even if creating that data wasn't the fastest op ever.
* result-set joins will need to be fast, because the time taken to generate the data is impacting on how quickly other components can avail of this data

So far, my impression is that we will end up making joins available in two forms: a built-in, friendly way, that may not be incredibly fast, and a user-defined (much like your last two examples) which relies on the user's knowledge of their application and can leverage indexes for the joins.

I would do left joins for now - see how it goes, and eventually implement both, if the left join works fine, what do you think?

One idea that's cooking in my mind is some kind of algorithm that keeps track of "distinct" (to use sql terminology) field values and record ids associated to those values, then executes an $or query on the second collection (with the $or values being those distinct values) and creates associations and generates the join. It's still kind-of forming in my mind so maybe it's not the clearest explanation, forgive me for that.

**Comment by jrhicks** [comment]
@assafweinberg @techfort Is Left-Right the way to think about it?  Top-Bottom makes sense to me.  Maybe Collated Result Set Makes more sense then Joined Result Set.  Please skip to minute 6:00 of  https://www.youtube.com/watch?v=lQLTiX93PL8

**Comment by assafweinberg** [comment]
@techfort - I agree, left outer joins are the easiest to get started with. I *think* we're on the same page on algorithm. We'd run over each table in the join to create the data structure that @jrhicks described above.

@jrhicks - I just used left and right since they have meaning in a relational context... same domain as "join" (vs collate/top/bottom).

**Comment by techfort** [comment]
@jrhicks @assafweinberg while developing Loki is one of my main activity during my "coding spare time", if you want to give a go to developing joins that is ok with me. Not trying to pass the buck here (in fact i think it would be a very interesting piece of development) but you have put already a lot of design work in it so I don't want to deprive you of the pleasure, if that is how you feel about it.

**Comment by jrhicks** [comment]
@techfort I can't stop thinking about dynamic views and applyMap.  If I understand Loki's dynamic maps, do they maintain themselfs effeciently after the original build?

**Comment by techfort** [comment]
they do. This happens in two stages though: the rebuilding of the index and the data itself. Data can be made available lazily by calling data(), or else you can force the rebuild of the data (manually in several ways) at every change in the index, either way dynamicview updates itself. Look at this for a live example: [plunker](http://plnkr.co/edit/rye2uw1rhncOmNxbMfnh?p=preview)

**Comment by obeliskos** [comment]
Our current approach is to evaluate documents as they are inserted, updated, and deleted (no evaluation needed).  This places a small load on those ops for the benefit of fast reads.

A more expensive operation is sorting which (our current approach) is to queue up an async sort phase.  If you insert 100 documents which affect the view then, when inserting is completed, you will encounter a sort. If no sorts have been applied, you should not encounter this performance hit.

I believe that is what you asking about? If not feel free to clarify.

**Comment by techfort** [comment]
@obeliskos thanks for clarifying!

**Comment by jrhicks** [comment]
@obeliskos @techfort - So applied finds and wheres happen on each CRUD operation.  And Loki throttles sort operations because its more intensive.  I think I got it.  How does it throttle them?

**Comment by jrhicks** [comment]
@techfort On that plunker you messaged to me .... each additional name is sorted correctly.  But the original inserts never get sorted.  Is this on purpose?

**Comment by jrhicks** [comment]
@obeliskos - How does LokiJS maintain the sort?  Is it performed in nLog(n)?

**Comment by obeliskos** [comment]
Throttling is currently just a flag and async 1 timeout.

The sort is currently using default array.sort() along with a custom comparer function.  The algorithm behind array.sort() differs among javascript vm implementations.  In the future we can always benchmark and pursue specific sorting algorithms.

**Comment by jrhicks** [comment]
@obeliskos I'm not proposing that it needs to be changed, I just wanted to understand.  I was trying to imagine the algorithm for maintaining sort one CRUD at a time, and for the life of me I can't imagine it.

**Comment by jrhicks** [comment]
@obeliskos @techfort I still need to work out my ideas, but LokiJS dynamic views are basically the same exact thing as Flux Stores.  Its way too much to explain in this thread, hopefully I can have blog post soon.

**Comment by techfort** [comment]
@jrhicks i'm really looking forward to that blog post!

**Comment by obeliskos** [comment]
Well if you do an update or insert on an object you could binary search to locate its position (if sort field is indexed) and then pay a price for array splice.  My general approach is to assume larger datasets will have more frequent updates and need this performance more that 1 off inserts. 

Flux looks pretty cool, I was noticing how a general dispatcher could be created for loki to do generic loki operations with payloads like { 'INSERT_OP', 'UserCollection', { username: 'test' }}.  Does that make sense or should dispatchers be domain specific?

**Comment by jrhicks** [comment]
@obeliskos great clarity on the 1-offs approach to maintaining sort with binary search indexing and array splices.

The best value of a dispatcher is a "traffic cop"!

Imagine you had many dynamic views that depended on each other, you would get a dependency graph!  The FLUX dispatcher lets all stores respond to any action and wait on each other, but the dispatcher can detect cycles and warn the programmer.

If you were to impliment applyMap to dynamic views, you would have to add to the API a way for the applyMap to register its dependencies ... so you could ensure they could respond to the Create/Update messages first.  At this point a FLUX dispatcher would be quite handy.

**Comment by jrhicks** [comment]
@obeliskos So lets say you have a Collection of uses and a Dynamic View on users.  Your ActionCreator may produce the action ({ 'INSERT_OP', 'UserCollection', { username: 'test' }}) and send it to both the collection and dynamic view.  The dispatcher would then help the dynamic view wait on the collection (if thats even needed, not sure).

**Comment by jrhicks** [comment]
What do Loki users user the most?  ResultSet, DynamicViews or MapReduce ... I know its apple and oranges, but I'm curious about the history of these.

**Comment by obeliskos** [comment]
Interesting... will require more research.  Dispatcher seems to employ several technologies which could be repurposed/reused such as event emitter and experimental protocols. 

Not sure about use, i'd say the easiest way to learn would be collection, resultset, dynamicview, mapreduce but every user has different requirements.

**Comment by jrhicks** [comment]
@techfort @obeliskos http://jrhicks.github.io/dynamic_loki_views_and_flux_stores/index.html

**Comment by obeliskos** [comment]
That's a great article, helps a great deal to figure out how loki might fit into such an architecture.

I'll have to set up a sample project today and see if I can simulate the full cycle / data flows.  

Do you see value in incorporating a map phase or function directly into the view pipeline?  We currently have mapReduce utility function on the dynamic view to translate existing 'results'.  If your map functions only serve to limit the fields returned we could add functionality to persist a named fieldmap such as ['firstname', 'lastname', 'age'] which you could name and retrieve possibly on a data call such as dynview.data('userAges').  If needed we could include as part of automatically maintained pipeline where those results could be referenced by dynview.mapDataArray[].

**Comment by techfort** [comment]
@jrhicks that's awesome! Off to study Flux properly. 
@obeliskos you inadvertently triggered an idea: if you don't pass anything to data() then you get a plain rebuild of the view. If you pass a string, you get a `Collection.prototype.extract` (basically a flat array of values of the property name passed to the function)  equivalent, and if you pass a function then your providing a logic yourself of what to return. What do you think?

**Comment by obeliskos** [comment]
@techfort sounds interesting maybe if typeof param is string run extract, if param is array maybe some kind of fieldmap (or extract supporting multiple properties), and if typeof param is function run map on it.

**Comment by techfort** [comment]
@obeliskos maybe that's a logic better implemented in extract() itself so data() only has to worry about it being a function or not.

**Comment by obeliskos** [comment]
Indeed however I would need to tell you what subset of docs to extract from for any of the extract functionality.

**Comment by techfort** [comment]
no prob, we can refactor extract() to be a statically available function
and `Collection.extract()` just a curry on the collection data, while
DynamicView would call a plain extract passing the rebuilt data. Does that
sound good?

**Comment by obeliskos** [comment]
Sure.  So I would pass you an array of objects to resolve within the data call or you would overload array such as data().extract()?  First is fine just wondering.

**Comment by techfort** [comment]
mmm - either really... maybe I would avoid appending yet another call to
data(), and just pass the array.

**Comment by obeliskos** [comment]
Great!

**Comment by jrhicks** [comment]
Glad you guise liked the post :)

**Comment by assafweinberg** [comment]
@jrhicks - I'm also a big fan of React.  I like the direction of using Loki as a dataStore!

@techfort - In the meantime, I'll go ahead and work on a PR for adding a mapJoin function to resultset.

**Comment by jrhicks** [comment]
@assafweinberg awesome :)

**Comment by techfort** [comment]
@assafweinberg ditto, awesome!

**Comment by assafweinberg** [comment]
Getting started by adding mapJoin(joinData, leftkey, rightKey, mapFunction) to resultset.  The joinData can be either an array or a Resultset object.  I have a few tests in place and it seems to work alright.  

Pretty straight forward except I had to create an anonymous collection inside the Resultset to house the results of the mapJoin so that the return value continues to be chainable.  Take a look and let me know if I'm not missing anything major and I can put in a PR.

https://github.com/assafweinberg/LokiJS/tree/fluent-joins

**Comment by jrhicks** [comment]
+1

**Comment by techfort** [comment]
awesome, all looks great to me!

**Comment by assafweinberg** [comment]
Excellent!  I think being able to map collections together in the middle of the chain could be really useful. Something to keep in mind is that the returned results are no longer live database objects.  We might consider making them immutable to avoid any confusion.

**Comment by jrhicks** [comment]
Can immutable be an option?  For example, I’m using some Flux libraries (alt) and they force everything immutable anyway.  It would be nice to not have to worry about immutable on top of immutable.

**Comment by assafweinberg** [comment]
I'd say if immutable gets in the way at all, we shouldn't do it.  Its just a nice safeguard to keep people from accidentally updating an object and expecting it to save back to the database.

**Comment by techfort** [comment]
@assafweinberg The PR is merged, I'm off to play with it! Thanks a lot. Will revert with feedback and - all going well - we can start benchmarking and talking other join types.

**Comment by obeliskos** [comment]
nifty!  Great work and +1 loki features!  

Shall this be documented or hidden ninja feature?

**Comment by techfort** [comment]
as soon as this is final it'll make it in the docs with frills bells,
vinterblot and human sacrifice to Odin!

**Comment by obeliskos** [comment]
Sounds Greaatt!!!!

**Comment by assafweinberg** [comment]
Thanks guys.  Test it out, and let me know if I've missed anything big with this pattern.

Stepping back, maybe we make the map function optional on both innerJoin and mapJoin. We can return a RethinkDB style result - i.e - [{left: obj, right:obj}] when no map function is passed in.  This might be useful for people who want live, updatable objects in their results.  

Then it might make sense to rename mapJoin to just leftJoin, so our API methods are the well understood "leftJoin" and "innerJoin", both with optional map functions.

**Comment by techfort** [comment]
@assafweinberg @obeliskos @jrhicks i am very satisfied with the state of joins at the moment. If you are all agreed with @assafweinberg 's suggestion to make the map function optional and rename mapJoin to leftJoin, I will go ahead with it and close this when it's done and thoroughly tested.
Ideally I'd like to publish 1.3 before the end of April, so let me know if this is a good plan for you guys.

**Comment by assafweinberg** [comment]
I believe I already made the map optional in my latest commit and renamed the function eqJoin to mimic the finctionality in rethinkDB. I'm flexible on the name, so whatever you guys think.

**Comment by techfort** [comment]
Ha, i didn't even notice that! Great, I'll go ahead with the renaming and
close when the next version is published.
Thanks!

**Comment by techfort** [comment]
Actually upon exploring rethinkDB (which I wasn't very familiar with) I think it's an amazing product, so in tribute to that I'll leave the name unchanged and close this. Thanks @assafweinberg !

**Comment by linonetwo** [comment]
Is this really supported?

I tried

```js
const db = new Loki('example.db');
const users = db.addCollection('users');
users.insert([
  { name: 'Thor', age: 35 },
  { name: 'Loki', age: 30 },
]);
const result = users.findOne({ age: { $gte: 35 } });
// DEBUG: console
console.log(`results`, result);
users.insert({ name: 'PP', age: 30, friend: result });

users.findAndUpdate({ name: 'Thor' }, (thor) => {
  thor.age = 36;
});

console.log('PP', users.findOne({ name: 'PP' }));
console.log('Thor', users.findOne({ name: 'Thor' }));
```

And it prints

```log
results {
  name: 'Thor',
  age: 35,
  meta: { revision: 0, created: 1607708984654, version: 0 },
  '$loki': 1
}
PP {
  name: 'PP',
  age: 30,
  friend: {
    name: 'Thor',
    age: 35,
    meta: { revision: 0, created: 1607708984654, version: 0 },
    '$loki': 1
  },
  meta: { revision: 0, created: 1607708984657, version: 0 },
  '$loki': 3
}
Thor {
  name: 'Thor',
  age: 36,
  meta: {
    revision: 1,
    created: 1607708984654,
    version: 0,
    updated: 1607708984657
  },
  '$loki': 1
}
```
